### DISCUSSION TITLE
Support for devices that show extra information in the model

### DISCUSSION THREAD

**Discussion by badnetmask** [andrew-codechimp/HA-Battery-Notes#1912] [Ideas]
The [SureHA integration](https://github.com/benleb/sureha) reports devices in the following format:
```
Manufacturer: Sure Petcare
Model: Feeder (XXXX-XXXXXXX)
```
Where `XXXX-XXXXXXX` is the device's [serial number](https://github.com/search?q=repo%3Abenleb%2Fsureha%20serial&type=code).
As I can see in your code, it only supports [exact matches on the model names](https://github.com/andrew-codechimp/HA-Battery-Notes/blob/main/custom_components/battery_notes/library.py#L117).
Is there a chance to add support for wildcard or regexes?

PS: I have manually added mine for the time being, so no rush on this.

**Comment by andrew-codechimp** [comment]
I am really trying to avoid regex's as I don't want to end up writing them for people with these devices that don't have a proper model and I'm not sure if a simple wildcard would cover all permutations.

The correct way would be for the integration not to have this in the model, or use the new model ID being introduced in HA 2024.8.
Maybe raise an issue with the integration saying model should be model only for this reason?

I am rewriting the library lookup at the moment which may make it easier to add pattern matching in future but I do wonder how much I should bend for badly designed integrations, I'll keep coming back to this one I'm sure.

**Reply by badnetmask** [reply]
I totally understand. Would you mind providing some links to the HA references, so I can open a bug report with them?

**Comment by andrew-codechimp** [comment]
And this got me thinking more, regex is overkill but an optional json schema addition for the library might be an easy way of handling this.  No promises when but I like this idea.

```
 "model_match_method": {
      "type": "string",
      "description": "The matching method for model, default exact.",
      "enum": ["startswith", "endswith", "contains"]
  }
```

**Reply by badnetmask** [reply]
Cool, that looks like a good compromise. Let me know if you need help testing!

**Reply by andrew-codechimp** [reply]
I have just put version 2.4.x into beta which adds support for this (you'll need to have chosen to add betas via HACS to see it)
If you want to add your device via a PR it should work, I've been using it with Apple device wildcards [here](https://github.com/andrew-codechimp/HA-Battery-Notes/blob/0433d3b3f3a63fa67f12b1256683a2fbf131b82e/custom_components/battery_notes/data/library.json#L667)
Or just supply me the battery type/quantity and I can add it.

**Reply by badnetmask** [reply]
Sure. Let's try that. Please see #1935.

**Reply by andrew-codechimp** [reply]
Thanks, just merged it.  A HA restart or maybe an integration reload will pull the library down.

**Reply by badnetmask** [reply]
Working as expected! Thanks!

**Reply by andrew-codechimp** [reply]
Great, thanks for confirming. Was good to finally cover this use case.
